### PR TITLE
4955: Added details for using SHA-256 and SHA-512 encryption with SASL.

### PR DESCRIPTION
**Pull request by msudberg**
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/5133)

### CODE REVIEW

**Comment by sploiselle** [comment]
Chatted with @msudberg about this and I'm going to do some testing on our end to see what we actually support before committing to a strategy to document these options, e.g. we might rely more heavily on referencing the `librdkafka` docs

**Comment by benesch** [comment]
Thanks very much for working on this, @msudberg. I thought it might be helpful if I drew up a table with the valid configurations as I understand them:

Parameter | Valid values | Details
----------|--------------|-------------
`security_protocol` | `plaintext`, `ssl`, `sasl_plaintext`, `sasl_ssl`
`sasl_mechanism` | `gssapi`, `plain`, `scram-sha-256`, `scram-sha-512`, `oauthbearer` | Required if `security_protocol` is `sasl_plain` or `sasl_ssl`.
`sasl_username`<br>`sasl_password` | - | Required if `sasl_mechanism` is `plain`, `scram-sha-256`, or `scram-sha-512`.
`sasl_kerberos_principal` | - | Required if `sasl_mechanism` is `gssapi`.
`sasl_kerberos_...` | - | Relevant only if `sasl_mechanism` is `gssapi`.

The biggest differences from what you've got are:
  * "scram" is specified as a `sasl_mechanism`, not a `security_protocol`
  * the Kerberos configs are only relevant if the `sasl_mechanism` is `gssapi` (which in turn is only relevant if the `security_protocol` is `sasl_plaintext` or `sasl_ssl`)

Not sure how to best fit all of that into the existing structure of the doc.

By the way, these configuration parameters generally directly correspond to librdkafka's, so looking at the upstream documentation can be helpful: https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md

**Comment by sploiselle** [comment]
Here are all of the configuration options we support:
https://github.com/MaterializeInc/materialize/blob/main/src/sql/src/kafka_util.rs#L119

You'll note that the upstream library we use whose documention you had provides
many, many more options than we permit, so I was probably premature in recommending
we consider just link to those docs.

@benesch correctly identified the valid values we support for the options he
mentioned.

The big change here is that we now support `scram` authentication, which we
didn't before. I didn't realize this was recently added, hence my confusion.
Like we chatted, this means that the references to only supporting GSSAPI need
to be changed.

I also believe that places where we mention `security_protocol` should be updated
to reflect the values @benesch mentioned, i.e. this is in correct:

```
- Materialize currently only supports:
  - `GSSAPI` as the `sasl_mechanisms`
  - `SASL_PLAINTEXT` as the `security_protocol`
```

It looks like we support all of the `librdkafka` options available.

The following Hugo partial gets used in the `CREATE SOURCE` docs and needs to be updated, as well:
`doc/user/layouts/partials/create-source/connector/kafka/with-options.html`

One last note: don't be afraid to totally restructure these tables. I might have
overly structured them in a way that made them less extensible than they needed
to be.

**Comment by benesch** [comment]
> The big change here is that we now support scram authentication, which we
didn't before. I didn't realize this was recently added, hence my confusion.

For what it's worth, we've supported SCRAM for.... well, since whenever we started allowing these options to be plumbed through. I didn't change anything to enable it.

**Comment by msudberg** [comment]
@sploiselle, @benesch - This is ready for another review. I do have one question: Do we need to specify options for oauthbearer anywhere?

File: doc/user/content/sql/create-sink.md [hunk 1]
```diff
@@ -53,7 +53,7 @@ Field | Value type | Description
 #### SSL `WITH` options
 
 Use the following options to connect Materialize to an SSL-encrypted Kafka
-cluster.
+cluster. For more detail, see [SSL-encrypted Kafka details](#ssl-encrypted-kafka-details).
```
Comment (sploiselle) [inline doc/user/content/sql/create-sink.md:56]: This link currently only exists at `/sql/create-source/avro-kafka/#ssl-encrypted-kafka-details`

File: doc/user/content/sql/create-source/avro-kafka.md [hunk 2]
```diff
@@ -107,13 +107,13 @@ This creates a source that...
   `localhost:9092`.
 - Is append-only.
 
-### Connecting to a Kafka broker using SASL PLAIN authentication
+### Connecting to a Kafka broker using SASL authentication
```
Comment (sploiselle) [inline doc/user/content/sql/create-source/avro-kafka.md:110]: There was a link to this in the release notes––would you mind updating that, as well? You can see any broken links by going to the bottom of the PR and clicking, next to **buildkite/tests/briefcase-lint-docs**, **Details**.

**Review by sploiselle — APPROVED** [review]
LGTM modulo the links; tysm for helping untangle all of this

**Comment by sploiselle** [comment]
@msudberg re `sasl_mechanisms=oauthbearer`, you need to specify `security_protocol` to either `sasl_ssl` or `sasl_plaintext`--good catch on that!

File: doc/user/content/sql/create-sink.md [hunk 3]
```diff
@@ -62,20 +62,22 @@ Field | Value | Description
...
 `ssl_ca_location` | `text` | The absolute path to the certificate authority (CA) certificate. Used for both SSL client and server authentication. If unspecified, uses the system's default CA certificates.
 
-#### Kerberos `WITH` options
+#### SASL `WITH` options
 
-Use the following options to connect Materialize to a Kerberized Kafka
-cluster.
+Use the following options to connect Materialize using SASL.
+
+For more detail about connecting to a Kerberized Kafka cluster,see [Kerberized Kafka details](#kerberized-kafka-details).
```
Comment (benesch) [inline doc/user/content/sql/create-sink.md:69]: copyedit nit: missing space after comma.

**Review by benesch — APPROVED** [review]

**Comment by escobera** [comment]
I'm trying this out but can't connect to my cluster, I'm not using kerberos and when I try to:

```sql
CREATE SOURCE processos_kafka FROM KAFKA BROKER 'kafka-stage.apps.tjdft.jus.br:443'
    TOPIC 'processos.public.processos'
    WITH (
        security_protocol = 'sasl_ssl',
        sasl_mechanism = 'scram-sha-512',
        sasl_username = 'prospeccao',
        sasl_password = '----',
        ssl_certificate_location = '/home/rafa/ca.crt'        
    )
    FORMAT BYTES;
```

I get `SQL Error [XX000]: ERROR: Can't seem to find local keytab cache. You must provide explicit sasl_kerberos_keytab or sasl_kerberos_kinit_cmd option.`

**Comment by sploiselle** [comment]
@escobera Ah, this an annoying––the option you want to use is `sasl_mechanisms = 'scram-sha-512'`; you just forgot the _s_ at the end of the option name. I'll file an issue to see what we can do to improve the ergonomics here.

**Comment by escobera** [comment]
Oh I see, I used https://github.com/MaterializeInc/materialize/pull/5133#issuecomment-749630170 as reference. 

Thanks a lot @sploiselle it is working now =)

**Comment by sploiselle** [comment]
@escobera Ah, makes sense! fwiw our [documentation](https://materialize.com/docs/sql/create-source/avro-kafka/) is usually a better source of truth. I also amended the comment to include the correct option name lest anyone else stumble across it in the future.

**Pull request merged**